In [64]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [65]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [66]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [67]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [68]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
numeric_columns = [
    "price_change_percentage_24h",
    "price_change_percentage_7d",
    "price_change_percentage_14d",
    "price_change_percentage_30d",
    "price_change_percentage_60d",
    "price_change_percentage_200d",
    "price_change_percentage_1y"
]
# Apply StandardScaler to the selected columns
scaled_data = StandardScaler().fit_transform(df_market_data[numeric_columns])

In [69]:
# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(scaled_data, columns=numeric_columns)

# Copy the crypto names from the original DataFrame
df_scaled["coin_id"] = df_market_data.index

# Set the coin_id column as the index
df_scaled.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
df_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [70]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [71]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    kmeans = KMeans(n_clusters=k, random_state=0)
    kmeans.fit(df_scaled)
    inertia.append(kmeans.inertia_)

print(inertia)

[287.0, 198.5718180979813, 145.63574772192862, 79.02243535120975, 66.21778006000338, 59.13907002537009, 48.40631449041361, 38.99358798927969, 34.40059545400234, 28.57360256096785, 23.890044355377952]


In [72]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k_values,
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

df_elbow.head()

k     inertia
0  1  287.000000
1  2  198.571818
2  3  145.635748
3  4   79.022435
4  5   66.217780

In [73]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_curve = df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve to Determine Optimal k",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2,
    width=800,
    height=400
)

# Display the Elbow Curve
elbow_curve


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [74]:
# Initialize the K-Means model using the best value for k
best_k = 4 
kmeans_model = KMeans(n_clusters=best_k, random_state=0)

In [75]:
# Fit the K-Means model using the scaled DataFrame
kmeans_model.fit(df_scaled)

KMeans(n_clusters=4, random_state=0)

In [76]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame

cluster_predictions = kmeans_model.predict(df_scaled)
# Print the resulting array of cluster values.
print(cluster_predictions)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [77]:
# Create a copy of the scaled DataFrame
df_scaled_copy = df_scaled.copy()

In [78]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_copy["cluster"] = cluster_predictions

# Display the copy of the scaled DataFrame
df_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        2

In [79]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scatter_plot = df_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="cluster",  
    hover_cols=["coin_id"], 
    title="Cryptocurrency Clusters Based on Price Changes",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    width=800,
    height=400
)

# Display the scatter plot
scatter_plot


:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [80]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [81]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(df_scaled_copy.drop(columns=["cluster"]))

df_pca = pd.DataFrame(
    pca_data,
    columns=["PCA1", "PCA2", "PCA3"],
    index=df_scaled_copy.index
)

# View the scaled PCA data
df_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

In [82]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
print(explained_variance)

total_explained_variance = explained_variance.sum()
print(total_explained_variance)

[0.3719856  0.34700813 0.17603793]
0.8950316570309842


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895 or 89.5%


In [83]:
# Create a new DataFrame with the PCA data.
df_pca = pd.DataFrame(
    pca_data,
    columns=["PCA1", "PCA2", "PCA3"],  
    index=df_scaled_copy.index          
)

# Copy the crypto names from the original scaled DataFrame
df_pca["coin_id"] = df_scaled_copy.index

# Set the coin_id column as the index
df_pca.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
df_pca.head()


PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [84]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))
k_values

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [85]:
# Create an empty list to store the inertia values
inertia = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    kmeans = KMeans(n_clusters=k, random_state=0)
    kmeans.fit(df_pca)
    inertia.append(kmeans.inertia_)

# Display the inertia values to verify
print(inertia)

[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159, 25.91545691045613, 21.890936274633233, 17.720698854473675, 13.838698708864571, 7.984354811896223]


In [86]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k_values,      
    "inertia": inertia  
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

df_elbow.head()

k     inertia
0  1  256.874086
1  2  168.811896
2  3  115.918077
3  4   49.665497
4  5   41.357314

In [87]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_curve = df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve to Determine Optimal k",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2,
    width=800,
    height=400
)

# Display the Elbow Curve
elbow_curve

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [90]:
# Initialize the K-Means model using the best value for k
k_model = KMeans(n_clusters=4, random_state=0)


In [92]:
# Fit the K-Means model using the PCA data
k_model.fit(df_pca)

KMeans(n_clusters=4, random_state=0)

In [98]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_clusters = k_model.predict(df_pca)

# Print the resulting array of cluster values.
pca_clusters

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [100]:
# Create a copy of the scaled PCA DataFrame
df_pca_copy = df_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca_copy["cluster"] = pca_clusters
# Display the copy of the scaled PCA DataFrame
df_pca_copy.head()

PCA1      PCA2      PCA3  cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        2
ethereum     -0.458261  0.458466  0.952877        2
tether       -0.433070 -0.168126 -0.641752        0
ripple       -0.471835 -0.222660 -0.479053        0
bitcoin-cash -1.157800  2.041209  1.859715        2

In [101]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scatter_plot = df_pca_copy.hvplot.scatter(
    x="PCA1",  
    y="PCA2",  
    c="cluster",  
    hover_cols=["coin_id"], 
    title="Cryptocurrency Clusters in PCA Space",
    xlabel="PCA1",
    ylabel="PCA2",
    width=800,
    height=400
)

# Display the scatter plot
scatter_plot

:Scatter   [PCA1]   (PCA2,cluster,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [103]:
# Composite plot to contrast the Elbow curves
inertia_original = []
for k in k_values:
    kmeans = KMeans(n_clusters=k, random_state=0)
    kmeans.fit(df_scaled)  
    inertia_original.append(kmeans.inertia_)

# 2. Compute the Elbow Curve With PCA 
inertia_pca = []
for k in k_values:
    kmeans = KMeans(n_clusters=k, random_state=0)
    kmeans.fit(df_pca)  
    inertia_pca.append(kmeans.inertia_)

# 3. Create DataFrames for both curves to plot with hvplot
df_original = pd.DataFrame({"k": k_values, "inertia": inertia_original})
df_pca = pd.DataFrame({"k": k_values, "inertia": inertia_pca})

# 4. Plot the Elbow Curves using hvplot
plot_original = df_original.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve: Original Scaled Data vs PCA-Transformed Data",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2,
    label="Original Scaled Data",
    width=800,
    height=400
)

plot_pca = df_pca.hvplot.line(
    x="k",
    y="inertia",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2,
    color="red",
    label="PCA-Transformed Data",
    width=800,
    height=400
)

# Scatter to represent markers
scatter_original = df_original.hvplot.scatter(
    x="k",
    y="inertia",
    color="blue",
    size=10,
    label="Original Scaled Data",
)

scatter_pca = df_pca.hvplot.scatter(
    x="k",
    y="inertia",
    color="red",
    size=10,
    label="PCA-Transformed Data",
)

# Combine both plots
composite_plot = plot_original * plot_pca * scatter_original * scatter_pca

# Display the composite plot
composite_plot


:Overlay
   .Curve.Original_Scaled_Data                :Curve   [k]   (inertia)
   .Curve.PCA_hyphen_minus_Transformed_Data   :Curve   [k]   (inertia)
   .Scatter.Original_Scaled_Data              :Scatter   [k]   (inertia)
   .Scatter.PCA_hyphen_minus_Transformed_Data :Scatter   [k]   (inertia)

In [108]:
scatter_pca = df_pca_copy.hvplot.scatter(
    x="PCA1",  # Use PCA1 for the x-axis
    y="PCA2",  # Use PCA2 for the y-axis
    c="cluster",  # Color points by cluster label
    hover_cols=["coin_id"],  # Show cryptocurrency names on hover
    title="Cryptocurrency Clusters in PCA Space",
    xlabel="PCA1",
    ylabel="PCA2",
    width=800,
    height=400
)

# 2. Scatter plot for clusters in the original feature space (price_change_percentage_24h, price_change_percentage_7d)
scatter_original = df_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h",  # Use price_change_percentage_24h for the x-axis
    y="price_change_percentage_7d",   # Use price_change_percentage_7d for the y-axis
    c="cluster",  # Color points by cluster label
    hover_cols=["coin_id"],  # Show cryptocurrency names on hover
    title="Cryptocurrency Clusters Based on Price Changes",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    width=800,
    height=400
)

# 3. Combine both scatter plots to compare the clusters
composite_plot = scatter_pca * scatter_original

# Display the composite plot
composite_plot


:Overlay
   .Scatter.I  :Scatter   [PCA1]   (PCA2,cluster,coin_id)
   .Scatter.II :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cluster,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The data is reduced to a simpler form, which makes the clustering process more efficient, as it focuses on the most important features. It also reduces and eliminates irrelevent or repetative features. 